crates/bevy_ecs/src/storage/blob_vec.rs

```diff
@@ -0,0 +1,406 @@
+    }
+
+    /// # Safety
+    /// `index` must be in bounds
+    /// Allows aliased mutable access to `index`'s data. Caller must ensure this does not happen
+    #[inline]
+    pub unsafe fn set_unchecked(&self, index: usize, value: *mut u8) {
+        let ptr = self.get_unchecked(index);
+        std::ptr::copy_nonoverlapping(value, ptr, self.item_layout.size());
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the value pointer points to data that matches self.item_layout
+    /// This [BlobVec] will take ownership of `value`'s data, which means the caller should ensure that data is not dropped early  
+    #[inline]
+    pub unsafe fn push(&mut self, value: *mut u8) {
+        let index = self.push_uninit();
+        self.set_unchecked(index, value);
+    }
+
+    /// increases the length by one (and grows the vec if needed) with uninitialized memory and returns the index
+    /// # Safety
+    /// the newly allocated space must be immediately populated with a valid value
+    #[inline]
+    pub unsafe fn push_uninit(&mut self) -> usize {
+        self.reserve(1);
+        let index = self.len;
+        self.len += 1;
+        index
+    }
+
+    /// # Safety
+    /// len must be <= capacity. if length is decreased, "out of bounds" items must be dropped. if length is increased,
+    /// newly added items must be immediately populated with valid values.
+    pub unsafe fn set_len(&mut self, len: usize) {
+        self.len = len;
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure the type matches self.item_layout
+    #[inline]
+    pub unsafe fn push_type<T>(&mut self, mut value: T) {
+        self.push((&mut value as *mut T).cast::<u8>());
+        std::mem::forget(value);
+    }
+
+    /// # Safety
+    /// It is the caller's responsibility to ensure that `index` is < self.len()
+    /// It is also the caller's responsibility to free the returned pointer.
+    /// Callers should _only_ access the returned pointer immediately after calling this function.
+    #[inline]
+    pub unsafe fn swap_remove_and_forget_unchecked(&mut self, index: usize) -> *mut u8 {
+        let last = self.len - 1;
+        let swap_scratch = (*self.swap_scratch.get()).as_ptr();
+        std::ptr::copy_nonoverlapping(
+            self.get_unchecked(index),
+            swap_scratch,
+            self.item_layout.size(),
+        );
+        std::ptr::copy(
+            self.get_unchecked(last),
+            self.get_unchecked(index),
+            self.item_layout.size(),
+        );
+        self.len -= 1;
+        swap_scratch
+    }
+
+    /// # Safety
+    /// index must be in-bounds
+    #[inline]
+    pub unsafe fn swap_remove_unchecked(&mut self, index: usize) {
+        let value = self.swap_remove_and_forget_unchecked(index);
+        (self.drop)(value)
+    }
+
+    /// # Safety
+    /// index must be in-bounds
+    pub unsafe fn swap_remove_type_unchecked<T>(&mut self, index: usize) -> T {
```

Comment:
```suggestion
    pub unsafe fn swap_remove_typed_unchecked<T>(&mut self, index: usize) -> T {
```

Reply:
Good call. Although now that I think about it these methods are only used in unit tests to make them easier to read. Its probably better to just remove them and rewrite the tests. That cuts down on the api surface / number of unsafe methods to verify/maintain and also ensures we're testing the methods actually used elsewhere.